extendr-api/tests/deserializer_tests.rs

```diff
@@ -0,0 +1,174 @@
+
+            #[derive(Deserialize, PartialEq, Debug)]
+            struct Str(String);
+            assert_eq!(from_robj::<Str>(&r!("xyz")), Ok(Str("xyz".into())));
+
+            #[derive(Deserialize, PartialEq, Debug)]
+            struct StrSlice<'a>(&'a str);
+            assert_eq!(from_robj::<StrSlice>(&r!("xyz")), Ok(StrSlice("xyz")));
+
+            // Structs are mapped to named lists.
+            #[derive(Deserialize, PartialEq, Debug)]
+            struct Struct { a: i32, b: f64 }
+            assert_eq!(from_robj::<Struct>(&r!(list!(a=1, b=2))), Ok(Struct{ a: 1, b: 2.0 }));
+
+            // Enums are mapped to named lists of lists.
+            #[derive(Deserialize, PartialEq, Debug)]
+            enum Enum {
+                Unit,
+                Newtype(u32),
+                Tuple(u32, u32),
+                Struct { a: u32 },
+            }
+
+            let j = r!("Unit");
+            let expected = Enum::Unit;
+            assert_eq!(expected, from_robj(&j).unwrap());
+
+            // If the name is wrong:
+            let j = r!("UnitX");
+            assert_eq!(from_robj::<Enum>(&j).is_err(), true);
+
+            let j = r!(list!(Newtype=1));
+            let expected = Enum::Newtype(1);
+            assert_eq!(expected, from_robj(&j).unwrap());
+
+            let j = r!(list!(Tuple=list!(1, 2)));
+            let expected = Enum::Tuple(1, 2);
+            assert_eq!(expected, from_robj(&j).unwrap());
+
+            let j = r!(list!(Struct=list!(a=1)));
+            let expected = Enum::Struct { a: 1 };
+            assert_eq!(expected, from_robj(&j).unwrap());
+
+            // Many things will generate a Robj.
+            // But note that the original Robj will not be copied verbatim.
+            // The Deserialize trait for Robj can also be used to generate
+            // JSON and other formats for Robj.
+            #[derive(Deserialize, PartialEq, Debug)]
+            struct ROBJ(Robj);
+            assert_eq!(from_robj::<ROBJ>(&r!(TRUE)), Ok(ROBJ(r!(TRUE))));
+            assert_eq!(from_robj::<ROBJ>(&r!(1)), Ok(ROBJ(r!(1))));
+            assert_eq!(from_robj::<ROBJ>(&r!(1.0)), Ok(ROBJ(r!(1.0))));
+            assert_eq!(from_robj::<ROBJ>(&r!("xyz")), Ok(ROBJ(r!("xyz"))));
+
+            // Sequences are always converted to lists.
+            assert_eq!(from_robj::<ROBJ>(&r!([TRUE, FALSE])), Ok(ROBJ(r!(list!(TRUE, FALSE)))));
+            assert_eq!(from_robj::<ROBJ>(&r!([1, 2])), Ok(ROBJ(r!(list!(1, 2)))));
+
+            // If you use a wrapper type, conversions are more specific.
+            #[derive(Deserialize, PartialEq, Debug)]
+            struct RIntegers(Integers);
+            assert_eq!(from_robj::<RIntegers>(&r!(1)), Ok(RIntegers(Integers::from_values([1]))));
+            assert_eq!(from_robj::<RIntegers>(&r!([1, 2])), Ok(RIntegers(Integers::from_values([1, 2]))));
+            assert_eq!(from_robj::<RIntegers>(&r!(1.0)).is_err(), true);
```

Comment:
I am not sure I understand why we can deserialize `r!(1.0)` as `Rint` but not as `Integers` (see assertion on line 36).